lightning/src/ln/channelmanager.rs

```diff
@@ -4915,14 +4981,62 @@ impl Readable for ClaimableHTLC {
 	}
 }
 
-impl_writeable_tlv_based_enum!(HTLCSource,
-	(0, OutboundRoute) => {
-		(0, session_priv, required),
-		(2, first_hop_htlc_msat, required),
-		(4, path, vec_type),
-	}, ;
-	(1, PreviousHopData)
-);
+impl Readable for HTLCSource {
+	fn read<R: Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let id: u8 = Readable::read(reader)?;
+		match id {
+			0 => {
+				let mut session_priv: ::util::ser::OptionDeserWrapper<SecretKey> = ::util::ser::OptionDeserWrapper(None);
+				let mut first_hop_htlc_msat = ::util::ser::OptionDeserWrapper(None);
+				let mut path = Some(Vec::new());
+				let mut mpp_id = None;
+				read_tlv_fields!(reader, {
+					(0, session_priv, required),
+					(2, first_hop_htlc_msat, required),
+					(4, path, vec_type),
+					(6, mpp_id, option),
+				});
+				if mpp_id.is_none() {
+					// For backwards compat, if there was no mpp_id written, use the session_priv bytes
+					// instead.
+					mpp_id = Some(MppId(*session_priv.0.unwrap().as_ref()));
```

Comment:
Seems the session_priv won't have been read yet if I do that

Reply:
`read_tlv_fields` should always overwrite the fields being read, at least if they're present in the TLV stream being read. It should otherwise leave them alone.

Reply:
Think I'm misunderstanding, is this the diff you're thinking of? Or are you thinking of initializing `mpp_id` with a random number instead of `session_priv`? 
```
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 4d9a965f..5792f4d2 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -4991,17 +4991,17 @@ impl Readable for HTLCSource {
                                let mut first_hop_htlc_msat: u64 = 0;
                                let mut path = Some(Vec::new());
                                let mut mpp_id = None;
+                               if mpp_id.is_none() {
+                                       // For backwards compat, if there was no mpp_id written, use the session_priv bytes
+                                       // instead.
+                                       mpp_id = Some(MppId(*session_priv.0.unwrap().as_ref()));
+                               }
                                read_tlv_fields!(reader, {
                                        (0, session_priv, required),
                                        (1, mpp_id, option),
                                        (2, first_hop_htlc_msat, required),
                                        (4, path, vec_type),
                                });
-                               if mpp_id.is_none() {
-                                       // For backwards compat, if there was no mpp_id written, use the session_priv bytes
-                                       // instead.
-                                       mpp_id = Some(MppId(*session_priv.0.unwrap().as_ref()));
-                               }
                                Ok(HTLCSource::OutboundRoute {
                                        session_priv: session_priv.0.unwrap(),
                                        first_hop_htlc_msat: first_hop_htlc_msat,

Reply:
Ohhhhhhhhhhhhhhhhhhhhhh, yes, you're totally right, that is the diff I was thinking of, and, yes, it is nonsensical.